### PR TITLE
Various LGTM fixes (file permissions, file check before open, etc.)

### PR DESCRIPTION
**Pull request by karlhto**

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a pull request! Maybe @olehermanse can review this?

**Comment by karlhto** [comment]
> This pull request **fixes 15 alerts** when merging [3f41943](https://github.com/cfengine/core/commit/3f41943f9e8c60cb76bf5ba045eaf6dee8cbfd21) into [35d777b](https://github.com/cfengine/core/commit/35d777b0fd54b574b42cf61c202bf077ed986e8f) - [view on LGTM.com](https://lgtm.com/projects/g/cfengine/core/rev/pr-5eb1360aae86e17df6a7c0a3b763afe558752937)
> 
> **fixed alerts:**
> 
> * 9 for File created without restricting permissions
> * 5 for Time-of-check time-of-use filesystem race condition
> * 1 for Uncontrolled data used in path expression

I think some of these are false. I think LGTM stops reporting on for instance "time of check, time of open" warnings when using non-standard open functions (like `safe_open`).

**Comment by karlhto** [comment]
I think I have _actually_ fixed all of the aforementioned instances of file checking before opening now.

**Comment by karlhto** [comment]
Split changes relevant to `cf-upgrade` into a new PR: https://github.com/cfengine/core/pull/3814

**Comment by karlhto** [comment]
@olehermanse @vpodzime This should be ready for review now. :)

**Comment by karlhto** [comment]
@vpodzime @olehermanse Fixed! Now it's ready for review.

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2697/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2697/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2697/

File: cf-execd/cf-execd-runner.c [hunk 1]
```diff
@@ -337,22 +337,25 @@ void LocalExec(const ExecConfig *config)
...
     }
     else
     {
         Log(LOG_LEVEL_VERBOSE, "No output");
         unlink(filename);
+        fclose(fp);
     }
+
+    Log(LOG_LEVEL_DEBUG, "Closing fp");
+    fclose(fp);
```
Comment (karlhto) [inline cf-execd/cf-execd-runner.c]: This isn't supposed to be like this, whoops.

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics, please

**Comment by karlhto** [comment]
> @vpodzime @olehermanse Fixed! Now it's ready for review.

Turns out I lied. Sorry!
Now it's ready, however.

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2698/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2698/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2698/

File: cf-execd/cf-execd-runner.c [hunk 2]
```diff
@@ -337,21 +337,21 @@ void LocalExec(const ExecConfig *config)
...
-    Log(LOG_LEVEL_DEBUG, "Closing fp");
-    fclose(fp);
-
     Log(LOG_LEVEL_VERBOSE,
         complete ? "Command is complete" : "Terminated command");
 
     if (count)
     {
+        Log(LOG_LEVEL_DEBUG, "Closing fp");
+        fclose(fp);
```
Comment (vpodzime) [inline cf-execd/cf-execd-runner.c]: Why is this moved?

Comment (karlhto) [reply cf-execd/cf-execd-runner.c]: Because I want to close the file _after_ using `unlink` in the `else` case.

Comment (karlhto) [reply cf-execd/cf-execd-runner.c]: Because I wanted to close the file _after_ unlinking it in the else case.

Comment (vpodzime) [reply cf-execd/cf-execd-runner.c]: Then I'd move it **after** the whole `if-else` block.

Comment (karlhto) [reply cf-execd/cf-execd-runner.c]: But the thing is, I need to close it before it is opened within `MailResult`. That was the reason `cf-execd` failed earlier.

A better solution might be to pass the `FILE`-pointer to MailResult.

**Comment by olehermanse** [comment]
> I think some of these are false. I think LGTM stops reporting on for instance "time of check, time of open" warnings when using non-standard open functions (like safe_open).

@karlhto yes, some of their queries have "hardcoded" function names like open, fopen, malloc, etc. Don't let this discourage you though. It's still good to fix the default perms, even if LGTM wrongly indicates that you also fixed the race condition. (In the future, I / we can modify the queries to include safe_fopen and safe_open).

File: cf-execd/cf-execd-runner.c (hunk 2 above)
Comment (olehermanse) [reply cf-execd/cf-execd-runner.c]: Makes sense.

> A better solution might be to pass the FILE-pointer to MailResult.

TODO.

File: cf-execd/cf-execd-runner.c [hunk 3]
```diff
@@ -609,6 +619,7 @@ static void MailResult(const ExecConfig *config, const char *file)
         if (CompareResultEqualOrFiltered(config, file, prev_file))
         {
             Log(LOG_LEVEL_VERBOSE, "Mail report: previous output is the same as current so do not mail it");
+            fclose(fp);
```
Comment (olehermanse) [inline cf-execd/cf-execd-runner.c]: This function is pretty huge, and you have to add fclose in many places. I've seen many cases of people forgetting to add a close or free in cases like this. A nice pattern is to put most of the code in a helper function, and then fclose after the function call. Keep in mind for the future.
